core/src/com/unciv/ui/screens/pickerscreens/ImprovementPickerScreen.kt

```diff
@@ -39,22 +39,25 @@ class ImprovementPickerScreen(
 
     private var selectedImprovement: TileImprovement? = null
     private val gameInfo = tile.tileMap.gameInfo
-    private val ruleSet = gameInfo.ruleset
+    private val ruleset = gameInfo.ruleset
     private val currentPlayerCiv = gameInfo.getCurrentPlayerCivilization()
     // Support for UniqueType.CreatesOneImprovement
     private val tileMarkedForCreatesOneImprovement = tile.isMarkedForCreatesOneImprovement()
 
     private fun getRequiredTechColumn(improvement: TileImprovement) =
-        ruleSet.technologies[improvement.techRequired]?.column?.columnNumber ?: -1
+        ruleset.technologies[improvement.techRequired]?.column?.columnNumber ?: -1
 
-    fun accept(improvement: TileImprovement?) {
+    fun accept(improvement: TileImprovement?, secondImprovement: TileImprovement? = null) {
         if (improvement == null || tileMarkedForCreatesOneImprovement) return
         if (improvement.name == Constants.cancelImprovementOrder) {
             tile.stopWorkingOnImprovement()
             // no onAccept() - Worker can stay selected
         } else {
-            if (improvement.name != tile.improvementInProgress)
+            if (improvement.name != tile.improvementInProgress) {
                 tile.startWorkingOnImprovement(improvement, currentPlayerCiv, unit)
+                if (secondImprovement != null)
+                    tile.queueImprovement(secondImprovement, currentPlayerCiv, unit)
+            }
             unit.action = null // this is to "wake up" the worker if it's sleeping
             onAccept()
```

Comment:
I think it would be better to think of this recursively. What if we wanted to build an upgraded improvement after the next improvement? The list could go on more than twice, so can we modify this to allow for longer improvement chains?

Reply:
Agree, if the improvement list is of arbitrary length then this function should recieve a vararg or list